### PR TITLE
Flatpak : fix lockfile path, and remove danging file

### PR DESCRIPTION
**Pull request by simpledork47**
## Description
It's a fix for running elementum within Flatpak 

lockfile path was resolved to addons/plugin.video.elementum/.lockfile whereas on disk it's userdata/addon_data/plugin.video.elementum/.lockfile

in the case of Flatpak, there's no dangling process : they are killed at the end, so it's safe to remove the file completely at start

(https://github.com/elgatito/plugin.video.elementum/issues/1058)

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Translation

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by antonsoroko** [comment]
@simpledork47 
2 things:

1. there is a small bug in your code.

if lockfile does not exist then it fails:
`[Errno 2] No such file or directory: '/home/user/.var/app/tv.kodi.Kodi/data/userdata/addon_data/plugin.video.elementum/.lockfile'`
you need to check if file exists before remove it, e.g. add `os.path.exists(lockfile) and` to beginning of your `if`.

2. i do not see how it should improve things. file location was wrong for python part, but at the same time if file was not found - then nothing will be done in python part (it will not try to kill old process) for any environment.
and with fix - again nothing will be done in python part for flatpak environment.
so basically nothing changed.

and in golang part code uses correct file and it will try to kill old process if it exists. see that `ensureSingleInstance` function i posted above.

but it does help.

after some digging - it looks like there is some bug in https://github.com/elgatito/elementum/tree/master/lockfile module that results in crash **only** in flatpak environment.
either we fail on "acquiring lock" or "on try to kill old process".

**Comment by simpledork47** [comment]
yeah it does help indeed

I've added a test prior to removal, as you suggested

**Comment by antonsoroko** [comment]
> It makes Flatpak Kodi a first-class citizen for desktop (or htpc) linux users

i dunno. alternatively you can use Kodi 21.1 from testing/upcoming release.
https://packages.ubuntu.com/search?keywords=kodi&searchon=names&suite=oracular&section=all
https://packages.debian.org/search?keywords=kodi&searchon=names&suite=testing&section=all

Flatpak is ok, but not perfect. So i think some people (including me) would prefer .deb packages to flatpak.

**Comment by antonsoroko** [comment]
@simpledork47 okay, i found issue.
if lock file does not exists (e.g. on first run) elementum writes to lockfile its own pid.
then on subsequent runs elementum read that pid and check if old process still alive, if so - it kills old process.

and now funny part. apparently Flatpaks assigns **same** PID to elementum almost every time. in my case in 30 runs i always had `50` as PID, and only once i had different number.

so elementum runs, reads pid file, finds process and commit suicide :-) because "old" pid in file is equal to "new" pid of elementum. b/c of that suicide many log messages are lost so i was not able to see it in log.

see https://github.com/elgatito/elementum/pull/136

also with this fix, your fix can be simplified - we just need to correct path to lockfile.

interesting that in python code we have check about "same PID", see https://github.com/elgatito/plugin.video.elementum/commit/3a6d39aecbf26b7f56350415ae1079207781c481

**Comment by simpledork47** [comment]
yeah Flatpak run an app within a fresh namespace, it's container isolation - not being able to see other processes
PID within this container haven't real system PID

On my test it was pid=54 for elementum within container ( pid inside lockfile ), I'd guess it start at 1 and increase, so for a specific machine it's quite guessable and somehow constant

While testing I used
`flatpak run --command=sh --devel  tv.kodi.Kodi
`
This wasn't usable for debugging : look like it's running in a fresh container (and not within the already running Kodi) so elementum ain't listed in ps

Great that the real issue was found and fixed

Of course it wasn't python side fault but it's way easier to fix this way imho
for a end user it's a bit difficult to recompile go part of elementum, whereas copying one .py file it's simple and easy

Real issue is why is there an outdated file with pid number sitting on disk ?
This file should be cleared when elementum binary is killed (SIGNAL handler) I'd think

**Comment by antonsoroko** [comment]
@simpledork47 I used `flatpak enter` for testing, it enters already running container.

**Comment by antonsoroko** [comment]
@simpledork47 

>Real issue is why is there an outdated file with pid number sitting on disk ?

so it was supposed to be removed via `defer` (at the end of main function, to to speak)
https://github.com/elgatito/elementum/blob/de4ac744eaa7e26131f4d4bea3da0d4ed01348f0/main.go#L184
https://github.com/elgatito/elementum/blob/de4ac744eaa7e26131f4d4bea3da0d4ed01348f0/lockfile/lockfile.go#L88

but shutdown was made as "coroutine" so it never reached the end of main, so it never was removed.
on classic environment (win/lin/android) this mistake was not the issue, but with flatpak it became visible.

i added a fix to https://github.com/elgatito/elementum/pull/136/files

**Comment by antonsoroko** [comment]
@simpledork47 can you please update PR to have only
`lockfile = os.path.join(get_addon_profile_dir(), ".lockfile")`
?

So elgatito will not have questions about other already unnecessary lines.

**Comment by simpledork47** [comment]
@antonsoroko done, and PR title has been updated

**Comment by antonsoroko** [comment]
Thanks. Now we wait for @elgatito to review and merge PR.

**Comment by elgatito** [comment]
@simpledork47 @antonsoroko Good finding and good work 🚀
